server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -237,28 +278,20 @@ def data_frame(self, filter, axis):
         }
         """
         try:
-            df = self.filter_dataframe(filter)
+            slice = self.filter_dataframe(filter)
         except KeyError as e:
             raise FilterError(f"Error parsing filter: {e}") from e
-        var_idx = df.var.index.tolist()
-        obs_idx = df.obs.index.tolist()
-        values = df.X
-        df_shape = df.shape
-        if df_shape[0] == 1:
-            values = values[None, :]
-        elif df_shape[1] == 1:
-            values = values[:, None]
+        # convert sparse slice to dense
+        X = slice._X.toarray() if sparse.issparse(slice._X) else slice._X
```

Comment:
I tend to avoid the one line if statements in python, but that's probably just me.

Reply:
Unless you feel strong, I think this is actually preferable in this case, and follows the ScanPy/AnnData examples.